tools/querytee/response_comparator.go

```diff
@@ -0,0 +1,208 @@
+type SamplesComparatorFunc func(expected, actual json.RawMessage) error
+
+type SamplesResponse struct {
+	Status string
+	Data   struct {
+		ResultType string
+		Result     json.RawMessage
+	}
+}
+
+var samplesComparator = map[string]SamplesComparatorFunc{
+	"matrix": compareMatrix,
+	"vector": compareVector,
+	"scalar": compareScalar,
+}
+
+// RegisterSamplesComparator helps with registering custom sample types
+func RegisterSamplesComparator(samplesType string, comparator SamplesComparatorFunc) {
+	samplesComparator[samplesType] = comparator
+}
+
+func CompareSamplesResponse(expectedResponse, actualResponse []byte) error {
+	var expected, actual SamplesResponse
+
+	err := json.Unmarshal(expectedResponse, &expected)
+	if err != nil {
+		return err
+	}
+
+	err = json.Unmarshal(actualResponse, &actual)
+	if err != nil {
+		return err
+	}
+
+	if expected.Status != actual.Status {
+		return fmt.Errorf("expected status %s but got %s", expected.Status, actual.Status)
+	}
+
+	if expected.Data.ResultType != actual.Data.ResultType {
+		return fmt.Errorf("expected resultType %s but got %s", expected.Data.ResultType, actual.Data.ResultType)
+	}
+
+	comparator, ok := samplesComparator[expected.Data.ResultType]
+	if !ok {
+		return fmt.Errorf("resultType %s not registered for comparison", expected.Data.ResultType)
+	}
+
+	return comparator(expected.Data.Result, actual.Data.Result)
+}
+
+type matrixResult struct {
+	Result []struct {
+		Metric map[string]string
+		Values [][]interface{}
+	}
+}
+
+func compareMatrix(expectedRaw, actualRaw json.RawMessage) error {
+	var expected, actual matrixResult
+
+	err := json.Unmarshal(expectedRaw, &expected.Result)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(actualRaw, &actual.Result)
+	if err != nil {
+		return err
+	}
+
+	if len(expected.Result) != len(actual.Result) {
+		return fmt.Errorf("expected %d metrics but got %d", len(expected.Result),
+			len(actual.Result))
+	}
+
+	metricStringToIndexMap := make(map[string]int, len(expected.Result))
+	for i, actualMetric := range actual.Result {
+		metricStringToIndexMap[labels.FromMap(actualMetric.Metric).String()] = i
+	}
+
+	for _, expectedMetric := range expected.Result {
+		actualMetricIndex, ok := metricStringToIndexMap[labels.FromMap(expectedMetric.Metric).String()]
+		if !ok {
+			return fmt.Errorf("expected metric %s missing from actual response", expectedMetric.Metric)
+		}
+
+		actualMetric := actual.Result[actualMetricIndex]
+		expectedMetricLen := len(expectedMetric.Values)
+		actualMetricLen := len(actualMetric.Values)
+
+		if expectedMetricLen != actualMetricLen {
+			err := fmt.Errorf("expected %d samples for metric %s but got %d", expectedMetricLen,
+				expectedMetric.Metric, actualMetricLen)
+			if expectedMetricLen > 0 && actualMetricLen > 0 {
+				level.Error(util.Logger).Log("msg", err.Error(), "oldest-expected-ts", int64(expectedMetric.Values[0][0].(float64)),
```

Comment:
Can we introduce a function to do interface{} -> float64 -> int64 conversion, and also do proper type-checks? Repeating conversion all over the place makes code harder to read.